### PR TITLE
Replace URLFetcher with SimpleURLLoader in AdsServiceImpl

### PR DESCRIPTION
**Pull request by simonhong**
Close https://github.com/brave/brave-browser/issues/4273

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests && npm run test-security`) on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_ads/browser/ads_service_impl.cc [hunk 1]
```diff
@@ -1154,58 +1156,86 @@ void AdsServiceImpl::URLRequest(
...
       ads::URLRequestMethod method,
       ads::URLRequestCallback callback) {
-  net::URLFetcher::RequestType request_type = URLMethodToRequestType(method);
-
-  net::URLFetcher* fetcher = net::URLFetcher::Create(
-      GURL(url), request_type, this).release();
-  fetcher->SetRequestContext(g_browser_process->system_request_context());
-  fetcher->SetAutomaticallyRetryOnNetworkChanges(kRetriesCountOnNetworkChange);
-
+  net::NetworkTrafficAnnotationTag traffic_annotation =
```
Comment (simonhong) [inline components/brave_ads/browser/ads_service_impl.cc]: @bridiver  Please fill this annotation. I don't know any background for this service.

**Comment by simonhong** [comment]
Is it sufficient to run existing test cases for verifying this change?
I don't know much about ads service. So, I can't do additional manual tests.

File: components/brave_ads/browser/ads_service_impl.cc (hunk 1 above)
Comment (bbondy) [reply components/brave_ads/browser/ads_service_impl.cc]: cc @NejcZdovc

File: components/brave_ads/browser/ads_service_impl.h [hunk 2]
```diff
@@ -239,7 +243,7 @@ class AdsServiceImpl : public AdsService,
   bat_ads::mojom::BatAdsServicePtr bat_ads_service_;
 
   NotificationInfoMap notification_ids_;
-  std::map<const net::URLFetcher*, ads::URLRequestCallback> fetchers_;
+  std::unordered_set<network::SimpleURLLoader*> url_loaders_;
```
Comment (iefremov) [inline components/brave_ads/browser/ads_service_impl.h]: nit: `Generally avoid std::unordered_set and std::unordered_map...` from `base/containers/README.md`

File: components/brave_ads/browser/ads_service_impl.cc [hunk 3]
```diff
@@ -1154,58 +1156,86 @@ void AdsServiceImpl::URLRequest(
...
+              [POLICY_NAME] {
+                  [POLICY_NAME]: ...
+              }
+            }
+            policy_exception_justification = "..."
+          }
+          comments: "..."
+        )");
+
+  auto request = std::make_unique<network::ResourceRequest>();
```
Comment (iefremov) [inline components/brave_ads/browser/ads_service_impl.cc]: probably we want
```
resource_request->load_flags =
      net::LOAD_DO_NOT_SEND_COOKIES | net::LOAD_DO_NOT_SAVE_COOKIES;
```

**Comment by iefremov** [comment]
LGTM!

File: components/brave_ads/browser/ads_service_impl.h (hunk 2 above)
Comment (simonhong) [reply components/brave_ads/browser/ads_service_impl.h]: Done.

File: components/brave_ads/browser/ads_service_impl.cc (hunk 3 above)
Comment (simonhong) [reply components/brave_ads/browser/ads_service_impl.cc]: Thanks for checking!
`request->allow_credentials = false;` is added here.
`DO_NOT_SAVE_COOKIES` and related flags are deprecated.

File: components/brave_ads/browser/ads_service_impl.cc [hunk 4]
```diff
@@ -260,6 +262,30 @@ bool SaveBundleStateOnFileTaskRunner(
   return false;
 }
 
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("ads_service_impl", R"(
```
Comment (simonhong) [inline components/brave_ads/browser/ads_service_impl.cc]: CC: @tmancey

Comment (NejcZdovc) [reply components/brave_ads/browser/ads_service_impl.cc]:
```
semantics {
        sender: "Brave Ads service"
        description:
          "This service is used to communicate with Brave servers "
          "to get and send information needed for ads to work."
        trigger:
          "Mostly they are triggered on intervals."
        data:
          "Ads catalog, confirmations and reporting."
        destination: WEBSITE
      }
      policy {
        cookies_allowed: NO
        setting:
          "You can enable or disable this feature by visiting brave://rewards."
        policy_exception_justification:
          "Not implemented."
}
```

Comment (simonhong) [reply components/brave_ads/browser/ads_service_impl.cc]: Updated. Thanks!

**Pull request merged**